%%% Preamble
\documentclass[paper=a4, fontsize=11pt]{scrartcl} % Article class of KOMA-script with 11pt font and a4 format

\usepackage[english]{babel} % English language/hyphenation
\usepackage[protrusion=true,expansion=true]{microtype} % Better typography
\usepackage{amsmath,amsfonts,amsthm} % Math packages
\usepackage[pdftex]{graphicx} % Enable pdflatex
%\usepackage{color,transparent} % If you use color and/or transparency
\usepackage[hang, small,labelfont=bf,up,textfont=it,up]{caption} % Custom captions under/above floats
\usepackage{epstopdf} % Converts .eps to .pdf
\usepackage{subfig} % Subfigures
\usepackage{booktabs} % Nicer tables
\usepackage{hyperref}

%%% Advanced verbatim environment
\usepackage{verbatim}
\usepackage{fancyvrb}
\DefineShortVerb{\|} % delimiter to display inline verbatim text


%%% Custom sectioning (sectsty package)
\usepackage{sectsty} % Custom sectioning (see below)
\allsectionsfont{ % Change font of al section commands
  \usefont{OT1}{bch}{b}{n} % bch-b-n: CharterBT-Bold font
% \hspace{15pt} % Uncomment for indentation
}

\sectionfont{ % Change font of \section command
 \usefont{OT1}{bch}{b}{n} % bch-b-n: CharterBT-Bold font
 \sectionrule{0pt}{0pt}{-5pt}{0.8pt} % Horizontal rule below section
}


%%% Custom headers/footers (fancyhdr package)
\usepackage{fancyhdr}
\pagestyle{fancyplain}
\fancyhead{} % No page header
%\fancyfoot[C]{\thepage} % Pagenumbering at center of footer
%\fancyfoot[R]{\small \texttt{}} % You can remove/edit this line 
\renewcommand{\headrulewidth}{0pt} % Remove header underlines
\renewcommand{\footrulewidth}{0pt} % Remove footer underlines
\setlength{\headheight}{12.0pt}

%%% Equation and float numbering
%\numberwithin{equation}{section}
% Equationnumbering: section.eq#
%\numberwithin{figure}{section}
% Figurenumbering: section.fig#
%\numberwithin{table}{section}
% Tablenumbering: section.tab#


%%% Title
\title{ \vspace{-1in} \usefont{OT1}{bch}{b}{n}
\huge \strut Quality Control in RNA-Seq Alignment \strut \\
\Large \bfseries \strut \strut
}
\author{ \usefont{OT1}{bch}{m}{n}
        Nima Hejazi\\ \usefont{OT1}{bch}{m}{n}
        Division of Biostatistics\\
        University of California, Berkeley\\ \usefont{OT1}{bch}{m}{n}
        \texttt{nh@nimahejazi.org}
}
\date{}

%%% Begin document
\begin{document}
\maketitle
\section{QC of RNA-Seq Reads (Post-Alignment)}
The present report concerns the process and results of performing quality
control (QC) post-alignment on RNA-Seq reads from the study described in the
{\href{http://www.nature.com/neuro/journal/v18/n1/abs/nn.3898.html}{``Developmental
regulation of human cortex transcription and its clinical relevance at base
resolution''}}, Jaffe \textit{et al.}, \textit{Nature Neuroscience}. In order to
perform basic quality control on RNA-Seq reads after the alignment procedure,
FastQC (v. 0.11.4) was used via the Galaxy computing platform. As the
re-analysis performed in this project was focused on only the \textit{paired-end
reads}, only these reads were used in the QC process. In all cases, more than
98\% of reads aligned successfully to the reference transcriptome. Most reads
aligned with exceptionally high quality (this is not usually the case in my past
experience), with a (rough) Phred score average of 33 or so. Based on a cursory
examination, there does not appear to be a significant trend between read
sequence alignment and either classes of samples (fetal versus adult) or sample
quality (as indicated by RIN). Based on the QC results, we conclude that the
alignment procedure was successful and adequate.

\subsection{The Results of Post-Alignment Quality Control}
The results of performing quality control on the samples is reported below:

\textbf{Adult samples:}
\begin{itemize}
    \item SRR1554535: 98.26\% of reads aligned concordantly, average quality
      score $\approx$ 33, reads with a score above 30 $\approx$ 77\%
    \item SRR1554536: 99.35\% of reads aligned concordantly, average quality
      score $\approx$ 34, reads with a score above 30 $\approx$ 84\%
    \item SRR1554539: 98.51\% of reads aligned concordantly, average quality
      score $\approx$ 35, reads with a score above 30 $\approx$ 92\%
\end{itemize}

\textbf{Fetal samples:}
\begin{itemize}
    \item SRR1554537: 98.42\% of reads aligned concordantly, average quality
      score $\approx$ 34.1, reads with a score above 30 $\approx$ 82\%
    \item SRR1554538: 98.51\% of reads aligned concordantly, average quality
      score $\approx$ 34.8, reads with a score above 30 $\approx$ 86\%
    \item SRR1554541: 98.59\% of reads aligned concordantly, average quality
      score $\approx$ 33.6, reads with a score above 30 $\approx$ 79\%
\end{itemize}

\end{document}
